src/detect-engine-build.c

```diff
@@ -1125,9 +1126,341 @@ static int RuleSetWhitelist(Signature *s)
     return wl;
 }
 
-int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, DetectPort **newhead,
-        uint32_t unique_groups, int (*CompareFunc)(DetectPort *, DetectPort *));
-int CreateGroupedPortListCmpCnt(DetectPort *a, DetectPort *b);
+static int CreateGroupedPortListCmpCnt(const DetectPort *a, const DetectPort *b)
+{
+    if (a->sh->init->score < b->sh->init->score) {
+        return 1;
+    } else if (a->sh->init->score > b->sh->init->score) {
+        return -1;
+    }
+
+    if (a->sh->init->sig_cnt < b->sh->init->sig_cnt) {
+        return 1;
+    } else if (a->sh->init->sig_cnt > b->sh->init->sig_cnt) {
+        return -1;
+    }
+
+    /* Hack to make the qsort output deterministic across platforms.
+     * This had to be done because the order of equal elements sorted
+     * by qsort is undeterministic and showed different output on BSD,
+     * MacOS and Windows. Sorting based on id makes it deterministic. */
+    if (a->sh->id < b->sh->id) {
+        return -1;
+    } else {
+        return 1;
+    }
+}
+
+static int SortCompare(const void *a, const void *b)
+{
+    const DetectPort *pa = *(const DetectPort **)a;
+    const DetectPort *pb = *(const DetectPort **)b;
+
+    return CreateGroupedPortListCmpCnt(pa, pb);
+}
+
+static inline void SortGroupList(uint32_t *groups, DetectPort **list,
+        int (*CompareFunc)(const DetectPort *, const DetectPort *))
+{
+    int cnt = 0;
+    for (DetectPort *x = *list; x != NULL; x = x->next) {
```

Comment:
did you mention somewhere this isn't needed as the list length is already known at this point?

Reply:
yeah but that was when there was a whole lot of other stuff I had done which is no longer there in the code